pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -190,6 +201,12 @@ func (nm *NodeManager) DiscoverNode(node *v1.Node) error {
 					klog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
 						node.Name, vm, res.vc, res.datacenter.Name())
 
+					// keep the zones discovered for the node
+					nodeZone := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
```

Comment:
What will be be behavior when the node object does not contain zones/region tags? Would this break volume provisioning, attach/detach?

Reply:
These maps will be empty in such case and hence shouldn't cause any issue. The existing regression tests without any zone labels were run with this change and they succeeded.

Reply:
Is there any information about how these nodes will have zone/region tags? Like who does? Will it come from vsphere.conf (on the node which has been deprecated) or from ESXi/vcenter?

Reply:
vsphere.conf specifies the labels section in this format which defines the tag category labels for zones and regions.

[Labels]
        zone = "k8s-io-zone"
        region = "k8s-io-region"

Now, a node can be assigned a zone/region by creating and assigning tags say zone1/EMEA under k8s-io-zone/k8s-io-region tag categories respectively in VC. A more detailed explanation is available here: https://github.com/kubernetes/kubernetes/issues/64021#issue-324351060 and https://github.com/kubernetes/kubernetes/issues/64021#issuecomment-407977363
The code logic is here: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L1324

Hope this clarifies. If this doesn't answer, please update.